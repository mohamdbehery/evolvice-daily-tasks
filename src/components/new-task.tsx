import React, { Component } from "react";
import Button from "../styled-components/button";
import TextArea from "../styled-components/textarea";
import Input from "../styled-components/input";
import Select from "../styled-components/select";
import { Helper } from "../helpers/helper";
import { Task } from "../types/Task";
import { AppBusiness } from "../helpers/app-business";
import { confirmAlert } from "react-confirm-alert";
import { Category } from "../types/Category";

interface NewTaskState {
    task: Task,
    categories: Category[]
}

export class NewTask extends Component<any, NewTaskState> {

    categoriesView: JSX.Element[] = [];
    helper = new Helper();
    appBusiness = new AppBusiness(this.props);

    constructor(props: any) {
        super(props);
        this.initState();
        this.name_Change = this.name_Change.bind(this);
        this.description_Change = this.description_Change.bind(this);
        this.category_Change = this.category_Change.bind(this);
        this.save_Click = this.save_Click.bind(this);
    }

    initState() {
        this.state = {
            task: {
                id: (this.props.match.params && this.props.match.params.id) ? +this.props.match.params.id : 0,
                categoryId: 0,
                categoryName: '',
                name: '',
                description: ''
            },
            categories: []
        };
    }

    componentDidMount() {
        this.appBusiness.GetCategories((categories: Category[]) => {
            this.setState({
                categories: categories
            });
        });

        if (this.state.task.id > 0) {
            this.appBusiness.GetTask(this.state.task.id, (task: Task) => {
                if (task) {
                    this.setState({ task });
                }
            })
        }
    }

    name_Change(event: any) {
        this.setState({
            task: {
                ...this.state.task,
                name: event.target.value
            }
        });
    }

    description_Change(event: any) {
        this.setState({
            task: {
                ...this.state.task,
                description: event.target.value
            }
        });
    }

    category_Change(event: any) {
        this.setState({
            task: {
                ...this.state.task,
                categoryId: +event.target.value
            }
        });
    }

    save_Click() {
        let newTask: Task = this.state.task;
        if (newTask && newTask.name && newTask.categoryId > 0 && newTask.description) {
            this.appBusiness.SaveTask(newTask, (status: boolean) => {
                if (status)
                    this.helper.redirectToPage('tasks');
            });
        }
        else {
            confirmAlert({
                title: 'Data required',
                message: 'Please enter all required fields',
                buttons: [
                    {
                        label: 'Ok',
                        onClick: () => true
                    }
                ]
            });
        }
    }

    hendleEnterKey(e: any) {
        if (e.key == 'Enter')
            this.save_Click();
    }

    render() {
        if (this.state.categories.length > 0) {
            this.categoriesView = this.state.categories.map((item: Category) => {
                return (<option value={item.id} key={item.id}>{item.name}</option>);
            });
        }
        return (
            <section className="section-property section-t8">
                <div className="container">
                    <div className="row">
                        <div className="col-md-12">
                            <div className="title-wrap d-flex justify-content-between">
                                <div className="title-box">
                                    <h2 className="title-a">Task Control</h2>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div className="row" onKeyPress={this.hendleEnterKey.bind(this)}>
                        <div className="col-md-6 mb-3">
                            <div className="form-group">
                                <Select value={this.state.task.categoryId} onChange={this.category_Change}>
                                    <option value={0} key={0}>Task category</option>
                                    {this.categoriesView}
                                </Select>
                                <div className="validation"></div>
                            </div>
                        </div>
                        <div className="col-md-6 mb-3">
                            <div className="form-group">
                                <Input type="text" name="title" placeholder="Task title" value={this.state.task.name} onChange={this.name_Change} />
                                <div className="validation"></div>
                            </div>
                        </div>
                        <div className="col-md-12 mb-3">
                            <div className="form-group">
                                <TextArea name="description" placeholder="Task description" value={this.state.task.description} onChange={this.description_Change}></TextArea>
                                <div className="validation"></div>
                            </div>
                        </div>
                        <div className="col-md-12">
                            <Button type="submit" onClick={this.save_Click}>Save</Button>
                            <Button onClick={() => this.helper.redirectToPage('tasks')}>Cancel</Button>
                        </div>
                    </div>
                </div>
            </section>
        )
    }
}

export default NewTask;