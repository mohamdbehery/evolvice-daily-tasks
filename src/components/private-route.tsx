import { Route, Redirect } from "react-router";
import React from "react";
import { Helper } from "../helpers/helper";

function PrivateRoute({ ...props }) {

    let helper = new Helper();
    let isAuthenticated = helper.getLoggedInUser() ? true : false;
    return (
        <Route
            path={props.path}
            render={({ location }) => {
                return isAuthenticated ? (props.children) : (<Redirect to={{ pathname: "/login", state: { from: location } }} />);
            }
            }
        />
    );
}

export default PrivateRoute;