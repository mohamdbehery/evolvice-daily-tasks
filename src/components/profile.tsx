import React, { Component } from "react";
import { Helper } from "../helpers/helper";
import { User } from "../types/User";

interface ProfileState {
    user: User
}

export class Profile extends Component<any, ProfileState> {

    helper = new Helper();
    constructor(props: any) {
        super(props);
        this.initState();        
    }

    initState(){
        this.state = {
            user:{
                id: 0,
                name: '',
                about: '',
                mail: '',
                username: '',
                password: '',
                address: '',
                fb: '',
                in: '',
                picture: '',
            }
        };
    }

    componentDidMount() {
        this.setState({
            user: this.helper.getLoggedInUser()
        });
    }

    render() {
        return (
            <section className="section-property section-t8">
                <div className="container">
                    <div className="row profile">
                        <div className="col-sm-6 col-md-4">
                            <img src={this.state.user.picture} alt="" className="img-profile" />
                        </div>
                        <div className="col-sm-6 col-md-8">
                            <h4>
                                {this.state.user.name}
                            </h4>
                            <div className="profile-address">
                                <i className="fa fa-location-arrow"></i> {this.state.user.address}
                            </div>
                            <p>
                                {this.state.user.about}
                            </p>
                            <p className="profile-links">
                                <i className="mail">
                                    {this.state.user.mail}
                                </i>
                                <br />
                                <i className="fa fa-linkedin"></i> {this.state.user.in}
                                <br />
                                <i className="fa fa-facebook"></i> {this.state.user.fb}
                            </p>
                        </div>
                    </div>
                </div>
            </section>
        )
    }
}

export default Profile;