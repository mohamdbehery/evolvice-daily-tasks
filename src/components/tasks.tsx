import React, { Component } from "react";
import { NavLink } from "react-router-dom";
import { Helper } from "../helpers/helper";
import { AppBusiness } from "../helpers/app-business";
import { Task } from "../types/Task";
import { Category } from "../types/Category";
import { confirmAlert } from "react-confirm-alert";

interface TasksState {
    id: number,
    tasks: Task[],
    filteredTasks: Task[],
    categories: Category[]
}

export class Tasks extends Component<any, TasksState> {

    appBusiness = new AppBusiness(this.props);
    helper = new Helper();
    tasksView: JSX.Element[] = [];
    categoriesView: JSX.Element[] = [];

    constructor(props: any) {
        super(props);
        this.initState();
        this.edit_Click = this.edit_Click.bind(this);
        this.goingToDelete = this.goingToDelete.bind(this);
        this.category_Change = this.category_Change.bind(this);
    }

    initState() {
        this.state = {
            id: 0,
            tasks: [],
            filteredTasks: [],
            categories: []
        };
    }

    componentDidMount() {
        this.appBusiness.GetUserTasks((response: TasksState) => {
            this.setState({
                tasks: response.tasks,
                filteredTasks: response.tasks,
                categories: response.categories
            })
        });
    }

    category_Change(event: any) {
        let categoryId: number = +event.target.value;
        let filteredTasks: Task[] = [];
        if (categoryId > 0) {
            filteredTasks = this.state.tasks.filter((task: Task) => {
                return task.categoryId === categoryId;
            });
        }
        else
            filteredTasks = this.state.tasks;

        this.setState({
            filteredTasks: filteredTasks
        });
    }

    edit_Click(event: any) {
        let id: number = +event.target.dataset.id;
        this.helper.redirectToPage('new-task/' + id);
    }

    delete_Click() {
        this.appBusiness.DeleteTask(this.state.id, (status: boolean, tasks: Task[]) => {
            if (status) {
                this.setState({
                    tasks: tasks,
                    filteredTasks: this.state.filteredTasks.filter((task: Task) => {
                        return task.id != this.state.id;
                    }),
                });
            }
        })
    }

    goingToDelete(event: any) {
        this.setState({ id: +event.target.dataset.id });
        confirmAlert({
            title: 'Confirmation',
            message: 'Are your sure to delete this?',
            buttons: [
                {
                    label: 'Yes',
                    onClick: () => { true; this.delete_Click() }
                },
                {
                    label: 'Cancel',
                    onClick: () => true
                }
            ]
        });
    }

    render() {
        this.tasksView = this.state.filteredTasks.map((task: Task) => {
            return (
                <div className="col-md-6" key={task.id}>
                    <div className="card card-body text-primary">
                        <h5 className="card-title">
                            <span className="highlighted">{task.categoryName}</span> {task.name}
                            <span className="admin card-settings">
                                <i className="fa fa-edit" data-id={task.id} onClick={this.edit_Click}></i>
                                <i className="fa fa-remove" data-id={task.id} onClick={this.goingToDelete}></i>
                            </span>
                        </h5>
                        <p className="card-text">{task.description}</p>
                    </div>
                </div>
            );
        });

        if (this.state.categories) {
            this.categoriesView = this.state.categories.map((category: Category) => {
                return (
                    <option value={category.id} key={category.id}>{category.name}</option>
                )
            })
        }
        return (
            <section className="section-property section-t8">
                <div className="container">
                    <div className="row">
                        <div className="col-md-12">
                            <div className="title-wrap d-flex justify-content-between">
                                <div className="title-box">
                                    <h2 className="title-a">Your Daily Tasks</h2>
                                </div>
                                <div className="title-link">
                                    <select className="custom-ddl" onChange={this.category_Change}>
                                        <option value={0} key={0}>All Categories</option>
                                        {this.categoriesView}
                                    </select>
                                    <div className="admin add-new">
                                        <NavLink className="nav-link add-new" to={'/new-task'}><i className="fa fa-plus"></i> Add new task</NavLink>
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                    <div className="row cards-container">
                        {this.tasksView}
                    </div>
                </div>
            </section>
        )
    }
}

export default Tasks;