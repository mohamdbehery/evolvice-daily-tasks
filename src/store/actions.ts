import { User } from "../types/User";

/**
 * all action types enum
 */
export enum ActionTypes {
    LOGIN = 'LOGIN',
    LOGOUT = 'LOGOUT'
} 

export interface ILoginAction{
    type: ActionTypes.LOGIN,
    payload: User
};
export interface ILogoutAction{
    type: ActionTypes.LOGOUT
};

export type UserActionTypes = ILoginAction | ILogoutAction;
export type AppActions = UserActionTypes;

export const loginAction = (user: User): AppActions => ({
    type: ActionTypes.LOGIN,
    payload: user
});
export const logoutAction = (): AppActions => ({
    type: ActionTypes.LOGOUT
});

