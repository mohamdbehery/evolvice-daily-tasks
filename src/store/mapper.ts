import { AppState } from "./store";
import { Dispatch } from "redux";
import { User } from "../types/User";
import { ActionTypes } from "./actions";

export const loginMapStateToProps = (state: AppState) => {
    return {
        login: state.login
    };
};

export const loginMapDispatchToProps = (dispatch: Dispatch) => {
    return {
        loginUser: (user: User) => {
            dispatch({
                type: ActionTypes.LOGIN,
                payload: user
            })
        },
        logoutUser: () => {
            dispatch({
                type: ActionTypes.LOGOUT,
            })
        }
    };
};