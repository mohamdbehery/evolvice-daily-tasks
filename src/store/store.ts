import { createStore, applyMiddleware, combineReducers, Store } from 'redux';
import { loginReducer } from './reducers';
import { AppActions } from './actions';

const rootReducer = combineReducers({
    login: loginReducer
});

export type AppState = ReturnType<typeof rootReducer>;
export const store = createStore(rootReducer, applyMiddleware(
    () => (next: any) => (action: AppActions)=> {
    console.log("Middleware Logger: " + action.type);
    next(action);
}));