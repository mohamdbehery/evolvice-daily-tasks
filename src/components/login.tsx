import React, { Component } from "react";
import { connect } from 'react-redux';
import Input from "../styled-components/input";
import Button from "../styled-components/button";
import { Helper } from "../helpers/helper";
import { AppBusiness } from "../helpers/app-business";
import { User } from "../types/User";
import { confirmAlert } from "react-confirm-alert";
import { loginMapStateToProps, loginMapDispatchToProps } from "../store/mapper";

interface LoginState {
    user: User
}
export class Login extends Component<any, LoginState> {

    appBusiness = new AppBusiness(this.props);
    helper = new Helper();

    constructor(props: any) {
        super(props);
        this.initState();
        this.username_Change = this.username_Change.bind(this);
        this.password_Change = this.password_Change.bind(this);
        this.login_Click = this.login_Click.bind(this);
    }

    initState() {
        this.state = {
            user: {
                username: '',
                password: ''
            }
        };
    }

    componentDidMount() {
        if (this.helper.getLoggedInUser())
            this.helper.redirectToPage('tasks');
    }

    login_Click() {
        let guest: User = this.state.user;
        if (guest && guest.username && guest.password) {
            this.appBusiness.Login(guest, (foundedUser: User) => {
                if (foundedUser) {
                    this.props.loginUser(foundedUser);
                    this.helper.setToLocalStorage("userData", foundedUser);
                    this.helper.redirectToPage('tasks');
                }
                else {
                    confirmAlert({
                        title: 'Invalid credentials',
                        message: 'Sorry, you have entered a wrong username or password',
                        buttons: [
                            {
                                label: 'Ok',
                                onClick: () => true
                            }
                        ]
                    });
                }
            });
        }
        else {
            confirmAlert({
                title: 'Data required',
                message: 'Please enter a valid username and password',
                buttons: [
                    {
                        label: 'Ok',
                        onClick: () => true
                    }
                ]
            });
        }
        if (this.props.login_Click)
            this.props.login_Click();
    }

    username_Change(event: any) {
        this.setState(
            {
                user:
                {
                    ...this.state.user,
                    username: event.target.value
                }
            });

        if (this.props.username_Change)
            this.props.username_Change();
    }

    password_Change(event: any) {
        this.setState(
            {
                user:
                {
                    ...this.state.user,
                    password: event.target.value
                }
            });
        if (this.props.password_Change)
            this.props.password_Change();
    }

    hendleEnterKey(e: any) {
        if (e.key == 'Enter')
            this.login_Click();
    }

    render() {
        return (
            <section className="section-property section-t8" data-testid="login-form">
                <div className="container">
                    <div className="login-form" onKeyPress={this.hendleEnterKey.bind(this)}>
                        <h2 className="text-center">Log in</h2>
                        <div className="form-group">
                            <Input data-testid="username" autoFocus type="text" className="form-control" placeholder="Username" value={this.state.user.username} onChange={this.username_Change} />
                        </div>
                        <div className="form-group">
                            <Input data-testid="password" type="password" className="form-control" placeholder="Password" value={this.state.user.password} onChange={this.password_Change} />
                        </div>
                        <div className="form-group">
                            <Button data-testid="login" type="submit" className="login-button" onClick={this.login_Click}>Log in</Button>
                        </div>
                    </div>
                </div>
            </section>
        )
    }
}

export default connect(loginMapStateToProps, loginMapDispatchToProps)(Login);