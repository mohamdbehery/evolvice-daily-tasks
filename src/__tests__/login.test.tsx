import React from "react";
import { render, fireEvent} from "@testing-library/react";
import { Login } from "../components/login";
import { store } from "../store/store";

describe("//- Login Component Test Cases -//", () => {
  test("Login form shot", async () => {
    const { findByTestId } = renderLogin();
    const loginForm = await findByTestId("login-form");
    const TextInputComponent = loginForm.shadowRoot;
    expect(TextInputComponent).toMatchSnapshot();
  });

  test("Login contains 2 inputs (username, password) and 1 login button", async () => {
    const { findByTestId } = renderLogin();
    const loginForm = await findByTestId("login-form");
    const inputs = loginForm.getElementsByTagName('input');
    const buttons = loginForm.getElementsByTagName('button');
    expect(inputs.length).toBe(2);
    expect(buttons.length).toBe(1);
  });

  test("Login check username input change", async () => {
    const usernameMoc = jest.fn();
    const { findByTestId } = renderLogin({ username_Change: usernameMoc });
    const username = await findByTestId("username");  
    fireEvent.change(username, { target: { value: "test" } });  
    expect(usernameMoc).toHaveBeenCalled();
  });

  
  test("Login check password input change", async () => {
    const passwordMoc = jest.fn();
    const { findByTestId } = renderLogin({ password_Change: passwordMoc });
    const password = await findByTestId("password");  
    fireEvent.change(password, { target: { value: "test" } });  
    expect(passwordMoc).toHaveBeenCalled();
  });

  test("Login check login button click", async () => {
    const loginMoc = jest.fn();
    const { findByTestId } = renderLogin({
      login_Click: loginMoc
    });
    const username = await findByTestId("username");
    const password = await findByTestId("password");
    const btnLogin = await findByTestId("login");
  
    fireEvent.change(username, { target: { value: "test" } });
    fireEvent.change(password, { target: { value: "password" } });
    fireEvent.click(btnLogin);
  
    expect(loginMoc).toHaveBeenCalled;
  });

  function renderLogin(props: Partial<any> = {}) {
    const defaultProps: any = {
      username_Change() {
        return;
      },
      password_Change() {
        return;
      },
      login_Click() {
        return;
      },
      store: store
    };
    return render(<Login {...defaultProps} {...props} />);
  }
});