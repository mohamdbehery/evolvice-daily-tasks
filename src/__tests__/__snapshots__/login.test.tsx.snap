// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`//- Login Component Test Cases -// Login form shot 1`] = `undefined`;
